File: deploy/relay/python/relay.py
#!/usr/bin/env python3
"""
OpenVort Relay — 企微消息中继服务（独立部署版）

部署在公网服务器上，接收企微回调消息，供本地/内网的 OpenVort 引擎拉取。
单文件，不依赖 openvort 包。

使用方法：
    1. pip install -r requirements.txt
    2. 修改下方配置（或设置环境变量）
    3. python relay.py
    4. 在企微后台填写回调 URL: http://your-server:8080/callback/wecom

环境变量（优先级高于下方默认值）：
    RELAY_WECOM_CORP_ID       企业ID
    RELAY_WECOM_APP_SECRET    应用Secret
    RELAY_WECOM_AGENT_ID      应用AgentId
    RELAY_WECOM_TOKEN         回调Token
    RELAY_WECOM_AES_KEY       回调EncodingAESKey
    RELAY_SECRET              Relay API 鉴权密钥（可选）
    RELAY_PORT                监听端口（默认 8080）
    RELAY_DB_PATH             SQLite 路径（默认 relay.db）
"""

import base64
import hashlib
import json
import os
import random
import sqlite3
import string
import struct
import time
import xml.etree.ElementTree as ET
from contextlib import asynccontextmanager

import httpx
import uvicorn
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from fastapi import FastAPI, Query, Request, Response
from fastapi.responses import JSONResponse

# ============================================================
# 配置（修改这里，或通过环境变量覆盖）
# ============================================================

WECOM_CORP_ID = os.getenv("RELAY_WECOM_CORP_ID", "")
WECOM_APP_SECRET = os.getenv("RELAY_WECOM_APP_SECRET", "")
WECOM_AGENT_ID = os.getenv("RELAY_WECOM_AGENT_ID", "")
WECOM_TOKEN = os.getenv("RELAY_WECOM_TOKEN", "")
WECOM_AES_KEY = os.getenv("RELAY_WECOM_AES_KEY", "")
WECOM_API_BASE = "https://qyapi.weixin.qq.com/cgi-bin"

RELAY_SECRET = os.getenv("RELAY_SECRET", "")  # API 鉴权密钥，为空则不鉴权
RELAY_PORT = int(os.getenv("RELAY_PORT", "8080"))
RELAY_DB_PATH = os.getenv("RELAY_DB_PATH", "relay.db")

# ============================================================
# 企微消息加解密
# ============================================================


class WeComCrypto:
    """企微回调消息签名验证 + AES 加解密"""

    def __init__(self, token: str, encoding_aes_key: str, corp_id: str):
        self.token = token
        self.corp_id = corp_id
        self.aes_key = base64.b64decode(encoding_aes_key + "=")

    def verify_signature(self, signature: str, timestamp: str, nonce: str, echostr: str = "") -> bool:
        items = sorted([self.token, timestamp, nonce, echostr])
        sha1 = hashlib.sha1("".join(items).encode()).hexdigest()
        return sha1 == signature

    def decrypt(self, encrypted: str) -> str:
        cipher_text = base64.b64decode(encrypted)
        iv = self.aes_key[:16]
        cipher = Cipher(algorithms.AES(self.aes_key), modes.CBC(iv))
        decryptor = cipher.decryptor()
        decrypted = decryptor.update(cipher_text) + decryptor.finalize()
        pad_len = decrypted[-1]
        content = decrypted[:-pad_len]
        msg_len = struct.unpack("!I", content[16:20])[0]
        msg = content[20:20 + msg_len].decode("utf-8")
        from_corp_id = content[20 + msg_len:].decode("utf-8")
        if from_corp_id != self.corp_id:
            raise ValueError(f"corp_id 不匹配: {from_corp_id} != {self.corp_id}")
        return msg

    def decrypt_callback(self, xml_text: str, msg_signature: str, timestamp: str, nonce: str) -> dict:
        root = ET.fromstring(xml_text)
        encrypt_node = root.find("Encrypt")
        if encrypt_node is None:
            raise ValueError("XML 中缺少 Encrypt 节点")
        encrypted = encrypt_node.text
        if not self.verify_signature(msg_signature, timestamp, nonce, encrypted):
            raise ValueError("签名验证失败")
        decrypted_xml = self.decrypt(encrypted)
        msg_root = ET.fromstring(decrypted_xml)
        return {child.tag: child.text or "" for child in msg_root}


# ============================================================
# SQLite 消息存储
# ============================================================


class RelayStore:
    """轻量消息存储"""

    def __init__(self, db_path: str):
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.execute("PRAGMA journal_mode=WAL")
        self.conn.execute("""
            CREATE TABLE IF NOT EXISTS messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                msg_id TEXT DEFAULT '',
                from_user TEXT DEFAULT '',
                msg_type TEXT DEFAULT 'text',
                content TEXT DEFAULT '',
                raw_data TEXT DEFAULT '{}',
                processed INTEGER DEFAULT 0,
                created_at REAL DEFAULT 0
            )
        """)
        self.conn.execute("CREATE INDEX IF NOT EXISTS idx_processed ON messages(processed)")
        self.conn.commit()

    def save(self, msg: dict) -> int:
        cur = self.conn.execute(
            "INSERT INTO messages (msg_id, from_user, msg_type, content, raw_data, created_at) VALUES (?,?,?,?,?,?)",
            (
                msg.get("MsgId", ""),
                msg.get("FromUserName", ""),
                msg.get("MsgType", "text"),
                msg.get("Content", ""),
                json.dumps(msg, ensure_ascii=False),
                float(msg.get("CreateTime", time.time())),
            ),
        )
        self.conn.commit()
        return cur.lastrowid

    def get_messages(self, since_id: int = 0, limit: int = 50) -> list[dict]:
        rows = self.conn.execute(
            "SELECT id, msg_id, from_user, msg_type, content, raw_data, processed, created_at "
            "FROM messages WHERE id > ? ORDER BY id ASC LIMIT ?",
            (since_id, limit),
        ).fetchall()
        return [
            {
                "id": r[0], "msg_id": r[1], "from_user": r[2], "msg_type": r[3],
                "content": r[4], "raw_data": json.loads(r[5]) if r[5] else {},
                "processed": bool(r[6]), "created_at": r[7],
            }
            for r in rows
        ]

    def mark_processed(self, msg_id: int):
        self.conn.execute("UPDATE messages SET processed = 1 WHERE id = ?", (msg_id,))
        self.conn.commit()

    def cleanup(self, max_age_hours: int = 72) -> int:
        cutoff = time.time() - max_age_hours * 3600
        cur = self.conn.execute("DELETE FROM messages WHERE created_at < ? AND processed = 1", (cutoff,))
        self.conn.commit()
        return cur.rowcount


# ============================================================
# 企微 Token 管理
# ============================================================

_token_cache = {"token": "", "expires": 0}


async def get_wecom_token() -> str:
    if _token_cache["token"] and time.time() < _token_cache["expires"] - 60:
        return _token_cache["token"]
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(
            f"{WECOM_API_BASE}/gettoken",
            params={"corpid": WECOM_CORP_ID, "corpsecret": WECOM_APP_SECRET},
        )
        data = resp.json()
    if data.get("errcode", 0) != 0:
        raise RuntimeError(f"获取 token 失败: {data}")
    _token_cache["token"] = data["access_token"]
    _token_cache["expires"] = time.time() + data.get("expires_in", 7200)
    return _token_cache["token"]


# ============================================================
# FastAPI 应用
# ============================================================

store: RelayStore | None = None
crypto: WeComCrypto | None = None


@asynccontextmanager
async def lifespan(app: FastAPI):
    global store, crypto
    store = RelayStore(RELAY_DB_PATH)
    if WECOM_TOKEN and WECOM_AES_KEY:
        crypto = WeComCrypto(WECOM_TOKEN, WECOM_AES_KEY, WECOM_CORP_ID)
    print(f"[Relay] 已启动 | DB: {RELAY_DB_PATH} | 加解密: {'✅' if crypto else '❌ 未配置'}")
    yield
    print("[Relay] 已关闭")


app = FastAPI(title="OpenVort Relay", version="0.1.0", lifespan=lifespan)


def check_auth(request: Request) -> bool:
    if not RELAY_SECRET:
        return True
    return request.headers.get("Authorization", "") == f"Bearer {RELAY_SECRET}"


# ---- 企微回调 ----


@app.get("/callback/wecom")
async def verify_url(
    msg_signature: str = Query(...),
    timestamp: str = Query(...),
    nonce: str = Query(...),
    echostr: str = Query(...),
):
    """企微回调 URL 验证"""
    if not crypto:
        return Response(content="crypto not configured", status_code=500)
    try:
        if crypto.verify_signature(msg_signature, timestamp, nonce, echostr):
            decrypted = crypto.decrypt(echostr)
            print(f"[Relay] URL 验证成功")
            return Response(content=decrypted, media_type="text/plain")
    except Exception as e:
        print(f"[Relay] URL 验证失败: {e}")
    return Response(content="verification failed", status_code=403)


@app.post("/callback/wecom")
async def receive_message(
    request: Request,
    msg_signature: str = Query(...),
    timestamp: str = Query(...),
    nonce: str = Query(...),
):
    """接收企微回调消息"""
    if not crypto:
        return Response(content="crypto not configured", status_code=500)
    try:
        body = await request.body()
        msg_dict = crypto.decrypt_callback(body.decode(), msg_signature, timestamp, nonce)
        msg_id = store.save(msg_dict)
        print(f"[Relay] 收到消息 #{msg_id}: {msg_dict.get('FromUserName', '')} -> {msg_dict.get('MsgType', '')} | {msg_dict.get('Content', '')[:50]}")
        return Response(content="success", media_type="text/plain")
    except Exception as e:
        print(f"[Relay] 回调处理失败: {e}")
        return Response(content="error", status_code=500)


# ---- Relay API ----


@app.get("/relay/messages")
async def get_messages(
    request: Request,
    since_id: int = Query(0),
    limit: int = Query(50),
):
    """拉取新消息"""
    if not check_auth(request):
        return JSONResponse({"error": "unauthorized"}, status_code=401)
    messages = store.get_messages(since_id=since_id, limit=limit)
    return JSONResponse({"messages": messages})


@app.post("/relay/messages/{msg_id}/ack")
async def ack_message(request: Request, msg_id: int):
    """标记消息已处理"""
    if not check_auth(request):
        return JSONResponse({"error": "unauthorized"}, status_code=401)
    store.mark_processed(msg_id)
    return JSONResponse({"ok": True})


@app.post("/relay/send")
async def send_message(request: Request):
    """代理发送消息到企微"""
    if not check_auth(request):
        return JSONResponse({"error": "unauthorized"}, status_code=401)

    body = await request.json()
    touser = body.get("touser", "")
    content = body.get("content", "")
    msg_type = body.get("msg_type", "text")

    if not touser or not content:
        return JSONResponse({"error": "touser and content required"}, status_code=400)

    try:
        token = await get_wecom_token()
        payload = {"touser": touser, "agentid": WECOM_AGENT_ID, "msgtype": msg_type}
        if msg_type == "markdown":
            payload["markdown"] = {"content": content}
        else:
            payload["text"] = {"content": content}

        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(f"{WECOM_API_BASE}/message/send?access_token={token}", json=payload)
            data = resp.json()

        if data.get("errcode", 0) != 0:
            print(f"[Relay] 发送失败: {data}")
            return JSONResponse({"error": data.get("errmsg", "unknown")}, status_code=502)

        print(f"[Relay] 已发送消息 -> {touser}")
        return JSONResponse({"ok": True})
    except Exception as e:
        print(f"[Relay] 发送异常: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@app.get("/relay/health")
async def health():
    """健康检查"""
    info = {
        "status": "ok",
        "crypto": bool(crypto),
        "corp_id": WECOM_CORP_ID[:6] + "***" if WECOM_CORP_ID else "",
        "db": RELAY_DB_PATH,
    }
    try:
        token = await get_wecom_token()
        info["wecom_api"] = bool(token)
    except Exception:
        info["wecom_api"] = False
    return JSONResponse(info)


@app.post("/relay/cleanup")
async def cleanup(request: Request):
    """清理过期消息"""
    if not check_auth(request):
        return JSONResponse({"error": "unauthorized"}, status_code=401)
    deleted = store.cleanup()
    return JSONResponse({"deleted": deleted})


# ============================================================
# 启动
# ============================================================

if __name__ == "__main__":
    if not WECOM_CORP_ID or not WECOM_APP_SECRET:
        print("❌ 请配置 RELAY_WECOM_CORP_ID 和 RELAY_WECOM_APP_SECRET")
        print("   可以修改 relay.py 顶部的配置，或设置环境变量")
        exit(1)

    print(f"""
╔══════════════════════════════════════════╗
║       OpenVort Relay Server v0.1         ║
╠══════════════════════════════════════════╣
║  端口: {RELAY_PORT:<33}║
║  数据库: {RELAY_DB_PATH:<31}║
║  加解密: {'已配置' if WECOM_TOKEN and WECOM_AES_KEY else '未配置（无法接收回调）':<25}║
║  鉴权: {'已启用' if RELAY_SECRET else '未启用':<27}║
╠══════════════════════════════════════════╣
║  回调地址:                               ║
║  http://your-domain:{RELAY_PORT}/callback/wecom{' ' * (5 - len(str(RELAY_PORT)))}║
║                                          ║
║  API 地址:                               ║
║  http://your-domain:{RELAY_PORT}/relay/{' ' * (12 - len(str(RELAY_PORT)))}║
╚══════════════════════════════════════════╝
""")
    uvicorn.run(app, host="0.0.0.0", port=RELAY_PORT, log_level="info")

File: deploy/relay/node/relay.js
#!/usr/bin/env node
/**
 * OpenVort Relay — 企微消息中继服务（Node.js 零依赖版）
 *
 * 部署在公网服务器上，接收企微回调消息，供本地/内网的 OpenVort 引擎拉取。
 * 零外部依赖，仅使用 Node.js 内置模块。
 *
 * 使用方法：
 *   1. 修改下方配置（或设置环境变量）
 *   2. node relay.js
 *   3. 在企微后台填写回调 URL: http://your-server:8080/callback/wecom
 *
 * 环境变量：
 *   RELAY_WECOM_CORP_ID       企业ID
 *   RELAY_WECOM_APP_SECRET    应用Secret
 *   RELAY_WECOM_AGENT_ID      应用AgentId
 *   RELAY_WECOM_TOKEN         回调Token
 *   RELAY_WECOM_AES_KEY       回调EncodingAESKey
 *   RELAY_SECRET              Relay API 鉴权密钥（可选）
 *   RELAY_PORT                监听端口（默认 8080）
 *   RELAY_DB_PATH             SQLite 路径（默认 relay.db）— 此版本使用 JSON 文件存储
 */

const http = require("http");
const https = require("https");
const crypto = require("crypto");
const fs = require("fs");
const path = require("path");
const { URL } = require("url");

// ============================================================
// 配置
// ============================================================

const CONFIG = {
  WECOM_CORP_ID: process.env.RELAY_WECOM_CORP_ID || "",
  WECOM_APP_SECRET: process.env.RELAY_WECOM_APP_SECRET || "",
  WECOM_AGENT_ID: process.env.RELAY_WECOM_AGENT_ID || "",
  WECOM_TOKEN: process.env.RELAY_WECOM_TOKEN || "",
  WECOM_AES_KEY: process.env.RELAY_WECOM_AES_KEY || "",
  WECOM_API_BASE: "https://qyapi.weixin.qq.com/cgi-bin",
  RELAY_SECRET: process.env.RELAY_SECRET || "",
  PORT: parseInt(process.env.RELAY_PORT || "8080"),
  DB_PATH: process.env.RELAY_DB_PATH || "relay_data.json",
};

// ============================================================
// 企微消息加解密
// ============================================================

class WeComCrypto {
  constructor(token, encodingAesKey, corpId) {
    this.token = token;
    this.corpId = corpId;
    this.aesKey = Buffer.from(encodingAesKey + "=", "base64");
    this.iv = this.aesKey.subarray(0, 16);
  }

  verifySignature(signature, timestamp, nonce, echostr = "") {
    const items = [this.token, timestamp, nonce, echostr].sort();
    const sha1 = crypto.createHash("sha1").update(items.join("")).digest("hex");
    return sha1 === signature;
  }

  decrypt(encrypted) {
    const decipher = crypto.createDecipheriv("aes-256-cbc", this.aesKey, this.iv);
    decipher.setAutoPadding(false);
    let decrypted = Buffer.concat([decipher.update(encrypted, "base64"), decipher.final()]);
    // 去除 PKCS7 填充
    const padLen = decrypted[decrypted.length - 1];
    decrypted = decrypted.subarray(0, decrypted.length - padLen);
    // 解析：16字节随机串 + 4字节消息长度 + 消息体 + corp_id
    const msgLen = decrypted.readUInt32BE(16);
    const msg = decrypted.subarray(20, 20 + msgLen).toString("utf-8");
    const fromCorpId = decrypted.subarray(20 + msgLen).toString("utf-8");
    if (fromCorpId !== this.corpId) {
      throw new Error(`corp_id 不匹配: ${fromCorpId} != ${this.corpId}`);
    }
    return msg;
  }

  decryptCallback(xmlText, msgSignature, timestamp, nonce) {
    // 提取 Encrypt 节点
    const match = xmlText.match(/<Encrypt><!\[CDATA\[(.*?)\]\]><\/Encrypt>/);
    if (!match) throw new Error("XML 中缺少 Encrypt 节点");
    const encrypted = match[1];
    if (!this.verifySignature(msgSignature, timestamp, nonce, encrypted)) {
      throw new Error("签名验证失败");
    }
    const decryptedXml = this.decrypt(encrypted);
    // 简单 XML 解析
    const result = {};
    const tagRegex = /<(\w+)><!\[CDATA\[([\s\S]*?)\]\]><\/\1>/g;
    const tagRegex2 = /<(\w+)>([\s\S]*?)<\/\1>/g;
    let m;
    while ((m = tagRegex.exec(decryptedXml)) !== null) result[m[1]] = m[2];
    while ((m = tagRegex2.exec(decryptedXml)) !== null) if (!result[m[1]]) result[m[1]] = m[2];
    return result;
  }
}

// ============================================================
// JSON 文件存储（零依赖，不用 SQLite）
// ============================================================

const PROCESSING_TIMEOUT_MS = 30000; // 处理中状态超时 30 秒

class RelayStore {
  constructor(dbPath) {
    this.dbPath = dbPath;
    this.data = { nextId: 1, messages: [] };
    this._load();

    // 启动超时清理定时器
    setInterval(() => this._releaseTimeout(), 5000);
  }

  _load() {
    try {
      if (fs.existsSync(this.dbPath)) {
        this.data = JSON.parse(fs.readFileSync(this.dbPath, "utf-8"));
      }
    } catch (e) {
      console.error(`[Relay] 加载数据失败: ${e.message}`);
    }
  }

  _save() {
    fs.writeFileSync(this.dbPath, JSON.stringify(this.data, null, 2));
  }

  // 释放超时的 processing 消息
  _releaseTimeout() {
    const now = Date.now();
    let changed = false;
    for (const msg of this.data.messages) {
      if (msg.status === "processing" && now - msg.processing_since > PROCESSING_TIMEOUT_MS) {
        msg.status = "pending";
        delete msg.processing_since;
        changed = true;
      }
    }
    if (changed) this._save();
  }

  save(msg) {
    const id = this.data.nextId++;
    this.data.messages.push({
      id,
      msg_id: msg.MsgId || "",
      from_user: msg.FromUserName || "",
      msg_type: msg.MsgType || "text",
      content: msg.Content || "",
      raw_data: msg,
      status: "pending", // pending | processing | processed
      created_at: parseFloat(msg.CreateTime) || Date.now() / 1000,
    });
    this._save();
    return id;
  }

  // 原子性获取并标记消息为处理中（防止多客户端重复消费）
  getAndMarkProcessing(sinceId = 0, limit = 50) {
    const now = Date.now();
    const messages = [];

    for (const msg of this.data.messages) {
      if (messages.length >= limit) break;
      if (msg.id <= sinceId) continue;
      if (msg.status === "processed") continue;
      if (msg.status === "pending") {
        // 原子性标记为 processing
        msg.status = "processing";
        msg.processing_since = now;
        messages.push(msg);
      } else if (msg.status === "processing") {
        // 超时的 processing 也视为可用
        if (now - (msg.processing_since || 0) > PROCESSING_TIMEOUT_MS) {
          msg.processing_since = now;
          messages.push(msg);
        }
      }
    }

    if (messages.length > 0) {
      this._save();
    }

    return messages;
  }

  // 确认处理完成
  markProcessed(msgId) {
    const msg = this.data.messages.find((m) => m.id === msgId);
    // 只有 processing 状态才能确认，防止重复 ack
    if (msg && msg.status === "processing") {
      msg.status = "processed";
      delete msg.processing_since;
      this._save();
      return true;
    }
    return false;
  }

  cleanup(maxAgeHours = 72) {
    const cutoff = Date.now() / 1000 - maxAgeHours * 3600;
    const before = this.data.messages.length;
    this.data.messages = this.data.messages.filter(
      (m) => !(m.status === "processed" && m.created_at < cutoff)
    );
    this._save();
    return before - this.data.messages.length;
  }

  // 调试用：获取所有消息（包括已消费的）
  getAllMessages(sinceId = 0, limit = 50, status = "all") {
    return this.data.messages
      .filter((m) => {
        if (m.id <= sinceId) return false;
        if (status !== "all" && m.status !== status) return false;
        return true;
      })
      .slice(-limit) // 返回最新的
      .reverse(); // 按时间倒序
  }

  // 调试用：重置消息状态
  resetMessage(msgId, newStatus = "pending") {
    const msg = this.data.messages.find((m) => m.id === msgId);
    if (msg) {
      msg.status = newStatus;
      if (newStatus === "pending") {
        delete msg.processing_since;
      }
      this._save();
      return true;
    }
    return false;
  }
}

// ============================================================
// 企微 Token 管理
// ============================================================

let tokenCache = { token: "", expires: 0 };

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try { resolve(JSON.parse(data)); } catch (e) { reject(e); }
      });
    }).on("error", reject);
  });
}

function httpsPost(url, body) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const postData = JSON.stringify(body);
    const options = {
      hostname: parsed.hostname,
      port: 443,
      path: parsed.pathname + parsed.search,
      method: "POST",
      headers: { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(postData) },
    };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try { resolve(JSON.parse(data)); } catch (e) { reject(e); }
      });
    });
    req.on("error", reject);
    req.write(postData);
    req.end();
  });
}

async function getWecomToken() {
  if (tokenCache.token && Date.now() / 1000 < tokenCache.expires - 60) {
    return tokenCache.token;
  }
  const url = `${CONFIG.WECOM_API_BASE}/gettoken?corpid=${CONFIG.WECOM_CORP_ID}&corpsecret=${CONFIG.WECOM_APP_SECRET}`;
  const data = await httpsGet(url);
  if (data.errcode) throw new Error(`获取 token 失败: ${JSON.stringify(data)}`);
  tokenCache.token = data.access_token;
  tokenCache.expires = Date.now() / 1000 + (data.expires_in || 7200);
  return tokenCache.token;
}

// ============================================================
// HTTP 服务
// ============================================================

const store = new RelayStore(CONFIG.DB_PATH);
const wecomCrypto =
  CONFIG.WECOM_TOKEN && CONFIG.WECOM_AES_KEY
    ? new WeComCrypto(CONFIG.WECOM_TOKEN, CONFIG.WECOM_AES_KEY, CONFIG.WECOM_CORP_ID)
    : null;

function checkAuth(req) {
  if (!CONFIG.RELAY_SECRET) return true;
  return (req.headers.authorization || "") === `Bearer ${CONFIG.RELAY_SECRET}`;
}

function parseQuery(url) {
  const params = {};
  const idx = url.indexOf("?");
  if (idx === -1) return params;
  url.substring(idx + 1).split("&").forEach((pair) => {
    const [k, v] = pair.split("=");
    params[decodeURIComponent(k)] = decodeURIComponent(v || "");
  });
  return params;
}

function readBody(req) {
  return new Promise((resolve) => {
    let body = "";
    req.on("data", (chunk) => (body += chunk));
    req.on("end", () => resolve(body));
  });
}

function json(res, data, status = 200) {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(data));
}

function text(res, content, status = 200) {
  res.writeHead(status, { "Content-Type": "text/plain" });
  res.end(content);
}

const server = http.createServer(async (req, res) => {
  const urlPath = req.url.split("?")[0];
  const query = parseQuery(req.url);

  try {
    // ---- 企微回调 ----

    if (urlPath === "/callback/wecom" && req.method === "GET") {
      if (!wecomCrypto) return text(res, "crypto not configured", 500);
      const { msg_signature, timestamp, nonce, echostr } = query;
      if (wecomCrypto.verifySignature(msg_signature, timestamp, nonce, echostr)) {
        const decrypted = wecomCrypto.decrypt(echostr);
        console.log("[Relay] URL 验证成功");
        return text(res, decrypted);
      }
      return text(res, "verification failed", 403);
    }

    if (urlPath === "/callback/wecom" && req.method === "POST") {
      if (!wecomCrypto) return text(res, "crypto not configured", 500);
      const body = await readBody(req);
      const { msg_signature, timestamp, nonce } = query;
      const msgDict = wecomCrypto.decryptCallback(body, msg_signature, timestamp, nonce);
      const msgId = store.save(msgDict);
      console.log(`[Relay] 收到消息 #${msgId}: ${msgDict.FromUserName || ""} -> ${msgDict.MsgType || ""} | ${(msgDict.Content || "").substring(0, 50)}`);
      return text(res, "success");
    }

    // ---- Relay API ----

    // 原子性获取消息并标记为处理中（防止多客户端重复消费）
    if (urlPath === "/relay/messages" && req.method === "GET") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const sinceId = parseInt(query.since_id || "0");
      const limit = parseInt(query.limit || "50");
      const messages = store.getAndMarkProcessing(sinceId, limit);
      return json(res, { messages });
    }

    // 确认消息处理完成
    if (urlPath.match(/^\/relay\/messages\/\d+\/ack$/) && req.method === "POST") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const msgId = parseInt(urlPath.split("/")[3]);
      const ok = store.markProcessed(msgId);
      return json(res, { ok });
    }

    if (urlPath === "/relay/send" && req.method === "POST") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const body = JSON.parse(await readBody(req));
      const { touser, content, msg_type = "text", media_id } = body;
      if (!touser) return json(res, { error: "touser required" }, 400);
      if (!content && msg_type !== "voice") return json(res, { error: "content required" }, 400);

      const token = await getWecomToken();
      const payload = { touser, agentid: CONFIG.WECOM_AGENT_ID, msgtype: msg_type };

      if (msg_type === "voice") {
        // 语音消息：需要 media_id
        if (!media_id) return json(res, { error: "media_id required for voice" }, 400);
        payload.voice = { media_id };
      } else if (msg_type === "markdown") {
        payload.markdown = { content };
      } else {
        payload.text = { content };
      }

      const data = await httpsPost(
        `${CONFIG.WECOM_API_BASE}/message/send?access_token=${token}`,
        payload
      );
      if (data.errcode) {
        console.log(`[Relay] 发送失败: ${JSON.stringify(data)}`);
        return json(res, { error: data.errmsg || "unknown" }, 502);
      }
      console.log(`[Relay] 已发送消息 -> ${touser} (${msg_type})`);
      return json(res, { ok: true });
    }

    if (urlPath === "/relay/health" && req.method === "GET") {
      const info = {
        status: "ok",
        crypto: !!wecomCrypto,
        corp_id: CONFIG.WECOM_CORP_ID ? CONFIG.WECOM_CORP_ID.substring(0, 6) + "***" : "",
        db: CONFIG.DB_PATH,
      };
      try {
        const token = await getWecomToken();
        info.wecom_api = !!token;
      } catch {
        info.wecom_api = false;
      }
      return json(res, info);
    }

    if (urlPath === "/relay/cleanup" && req.method === "POST") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const deleted = store.cleanup();
      return json(res, { deleted });
    }

    // 调试接口：获取所有消息（包括已消费的）
    if (urlPath === "/relay/messages/all" && req.method === "GET") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const sinceId = parseInt(query.since_id || "0");
      const limit = parseInt(query.limit || "50");
      const status = query.status || "all"; // pending, processing, processed, all
      const messages = store.getAllMessages(sinceId, limit, status);
      return json(res, { messages, total: store.data.messages.length });
    }

    // 调试接口：重置消息状态
    if (urlPath === "/relay/messages/reset" && req.method === "POST") {
      if (!checkAuth(req)) return json(res, { error: "unauthorized" }, 401);
      const { msg_id, status } = JSON.parse(await readBody(req) || "{}");
      const ok = store.resetMessage(msg_id, status);
      return json(res, { ok });
    }

    // 404
    json(res, { error: "not found" }, 404);
  } catch (e) {
    console.error(`[Relay] 错误: ${e.message}`);
    json(res, { error: e.message }, 500);
  }
});

// ============================================================
// 启动
// ============================================================

if (!CONFIG.WECOM_CORP_ID || !CONFIG.WECOM_APP_SECRET) {
  console.error("❌ 请配置 RELAY_WECOM_CORP_ID 和 RELAY_WECOM_APP_SECRET");
  console.error("   可以修改 relay.js 顶部的配置，或设置环境变量");
  process.exit(1);
}

console.log(`
╔══════════════════════════════════════════╗
║     OpenVort Relay Server v0.1 (Node)    ║
╠══════════════════════════════════════════╣
║  端口: ${String(CONFIG.PORT).padEnd(33)}║
║  存储: ${CONFIG.DB_PATH.padEnd(33)}║
║  加解密: ${(wecomCrypto ? "已配置" : "未配置（无法接收回调）").padEnd(25)}║
║  鉴权: ${(CONFIG.RELAY_SECRET ? "已启用" : "未启用").padEnd(27)}║
║  依赖: 零（纯 Node.js 内置模块）        ║
╠══════════════════════════════════════════╣
║  回调地址:                               ║
║  http://your-domain:${CONFIG.PORT}/callback/wecom  ║
║                                          ║
║  API 地址:                               ║
║  http://your-domain:${CONFIG.PORT}/relay/          ║
╚══════════════════════════════════════════╝
`);

server.listen(CONFIG.PORT, "0.0.0.0", () => {
  console.log(`[Relay] 已启动，监听 0.0.0.0:${CONFIG.PORT}`);
});
